### PR TITLE
DMP-843 Frontend Signin Integration - loginOrRefresh

### PR DESCRIPTION
**Pull request by DavidTMann**
This PR contains changes for;
DMP-843 Frontend internal login
DMP-801 Frontend internal handleoauth
DMP-802 Frontend internal logout callback

This could only be tested e2e locally against local darts-api instance due to mismatch in redirect uri domains. yarn dev:darts-api-stg will not work on internal user login, a separate ticket will be raised to accommodate this.

**Does this PR introduce a breaking change?** (check one with "x")

No breaking changes as long as signed in with external user. Currently internal user login is not 100% working against darts-api due to DMP-946.
```
[ ] Yes
[ X] No
```

### CODE REVIEW

File: server/controllers/authentication.ts [hunk 1]
```diff
@@ -68,11 +78,13 @@ function postAuthCallback(): (req: Request, res: Response, next: NextFunction) =
       }
     }
     try {
-      const result = await axios.post<SecurityToken>(EXTERNAL_USER_CALLBACK, req.body, {
+      let url;
+      type === 'internal' ? (url = INTERNAL_USER_CALLBACK) : (url = EXTERNAL_USER_CALLBACK);
```
Comment (davet1985) [inline server/controllers/authentication.ts:82]:
```suggestion
      const url = type === 'internal' ? INTERNAL_USER_CALLBACK) : EXTERNAL_USER_CALLBACK;
```

File: server/controllers/authentication.ts [hunk 2]
```diff
@@ -55,7 +63,9 @@ async function handleResetPassword(req: Request): Promise<string> {
   throw new Error(req.body);
 }
 
-function postAuthCallback(): (req: Request, res: Response, next: NextFunction) => Promise<void> {
+function postAuthCallback(
+  type: 'internal' | 'external'
+): (req: Request, res: Response, next: NextFunction) => Promise<void> {
   return async (req: Request, res: Response, next: NextFunction) => {
     console.log('postAuthCallback', req.body);
```
Comment (davet1985) [inline server/controllers/authentication.ts:70]: I know you didn't add it but could you remove?

File: server/controllers/authentication.ts [hunk 3]
```diff
@@ -99,12 +111,16 @@ function getLogout(): (_: Request, res: Response, next: NextFunction) => Promise
...
-        {
-          headers: { Authorization: `Bearer ${accessToken}` },
-        }
-      );
+      let logoutUrl: string;
+      if (req.session.userType === 'internal') {
+        logoutUrl = INTERNAL_USER_LOGOUT;
+      } else {
+        logoutUrl = EXTERNAL_USER_LOGOUT;
+      }
```
Comment (davet1985) [inline server/controllers/authentication.ts:119]:
```suggestion
      const logoutUrl = req.session.userType === 'internal' ? INTERNAL_USER_LOGOUT : EXTERNAL_USER_LOGOUT;
```

File: darts-api-stub/views/internal-login.ejs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    }
+  </style>
+  <body>
+    <h1>Stub login page</h1>
+    <form method="post" action="http://localhost:3000/auth/internal/callback">
+      <input type="hidden" name="code" value="fake-openid-code" />
+      <input id="login" type="submit" value="Login" />
+    </form>
+  </body>
+</html>
```
Comment (davet1985) [inline darts-api-stub/views/internal-login.ejs:19]: Could we use the same view as external with a param passed in for the callback?

**Review by davet1985 — COMMENTED** [review]
Couple of minor comment

File: darts-api-stub/authentication/external.js [hunk 5]
```diff
@@ -38,7 +38,7 @@ router.get('/logout', (_, res) => {
 });
 
 router.get('/login', (_, res) => {
-  res.render('external-login');
+  res.render('external-login', { callbackUrl: 'http://localhost:3000/auth/internal/callback' });
```
Comment (davet1985) [inline darts-api-stub/authentication/external.js:41]: Shouldn't this be the external callback?

Comment (DavidTMann) [reply darts-api-stub/authentication/external.js]: 🤦

**Review by davet1985 — APPROVED** [review]

**Pull request merged**